legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/notebook/PureBookLSPGrammarExtension.java

```diff
@@ -0,0 +1,323 @@
+        return globalState.getProperty(documentState.getDocumentId() + COMPILE_RESULT_KEY, () -> tryCompile(sectionState));
+    }
+
+    private CompletableFuture<? extends LambdaFunction<?>> tryCompile(SectionState sectionState)
+    {
+        return this.parse(sectionState).thenApply(x ->
+                {
+                    CompileResult compileResult = this.pureGrammarExtension.getCompileResult(sectionState);
+                    PureModel pureModel = compileResult.getPureModel();
+                    if (pureModel == null)
+                    {
+                        throw compileResult.getEngineException();
+                    }
+                    return HelperValueSpecificationBuilder.buildLambda(x, pureModel.getContext());
+                }
+                // when we complete compiling, trigger plan generation on the background to improve user experience...
+        ).whenCompleteAsync((l, e) -> this.generatePlan(sectionState), sectionState.getDocumentState().getGlobalState().getForkJoinPool());
+    }
+
+    @Override
+    public Iterable<? extends LegendDiagnostic> getDiagnostics(SectionState sectionState)
+    {
+        DocumentState documentState = sectionState.getDocumentState();
+        TextLocation sectionLocation = TextLocation.newTextSource(documentState.getDocumentId(), sectionState.getSection().getTextInterval());
+        if (!(documentState instanceof NotebookDocumentState))
+        {
+            return List.of(LegendDiagnostic.newDiagnostic(sectionLocation, "###purebook should not be use outside of Purebooks", LegendDiagnostic.Kind.Error, LegendDiagnostic.Source.Parser));
+        }
+
+        if (sectionState.getSection().getFullText().isEmpty())
+        {
+            return List.of();
+        }
+
+        try
+        {
+            try
+            {
+                this.compile(sectionState).get(30, TimeUnit.SECONDS);
+            }
+            catch (CompletionException | ExecutionException e)
+            {
+                throw e.getCause();
+            }
+        }
+        catch (EngineException e)
+        {
+            SourceInformation sourceInfo = e.getSourceInformation();
+            TextLocation location;
+            if (SourceInformationUtil.isValidSourceInfo(sourceInfo))
+            {
+                location = SourceInformationUtil.toLocation(sourceInfo);
+            }
+            else
+            {
+                location = sectionLocation;
+            }
+
+            LegendDiagnostic.Source source = e.getErrorType() == EngineErrorType.PARSER ? LegendDiagnostic.Source.Parser : LegendDiagnostic.Source.Compiler;
+
+            return List.of(LegendDiagnostic.newDiagnostic(location, e.getMessage(), LegendDiagnostic.Kind.Error, source));
+        }
+        catch (Throwable t)
+        {
+            return List.of(LegendDiagnostic.newDiagnostic(sectionLocation, t.getMessage(), LegendDiagnostic.Kind.Error, LegendDiagnostic.Source.Compiler));
+        }
```

Comment:
If it's not an EngineException, should it be returned as a diagnostic? Isn't it more likely to be indicative of a problem with the compiler than an issue with the user's code?

I would recommend logging the exception in this case, rather than returning the message as part of a diagnostic. After all, it might be something like a NullPointerException where there's no message. Or it could even be an Error (like an OutOfMemoryError) that indicates something going wrong with the JVM.

Reply:
The problem is that if something is going wrong on LSP and we don't send diagnostics, everything will look like is fine for the user.  I saw a non EngineException coming from one of the compilers, no diagnostic but then execution and other features don't work either.   Hence, is confusing to users...

I took the more conservative (or aggressive from some POV) to send broad diagnostics if we have some non Engine exception failures so users know that something is not working.  

You are right that catastrophic errors like OOM can be confusing, but at that point, we likely need to work on how to tell VSCode about these.